lib/block_view.go

```diff
@@ -4826,7 +5378,579 @@ func (bav *UtxoView) _connectUpdateProfile(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCreateNFT: ")
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorCreateNFTRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
+		// public key.
+	}
+
+	// Since issuing N copies of an NFT multiplies the downstream processing overhead by N,
+	// we charge a fee for each additional copy minted.
+	nftFee := txMeta.NumCopies * bav.GlobalParamsEntry.CreateNFTFeeNanos
+
+	// Ensure that the transaction covers the NFT fee.
+	totalOutput += nftFee
+	if totalInput < totalOutput {
+		return 0, 0, nil, RuleErrorCreateNFTWithInsufficientFunds
+	}
+
+	// Save a copy of the post entry so that we can safely modify it.
+	prevPostEntry := &PostEntry{}
+	*prevPostEntry = *postEntry
+
+	// Update and save the post entry.
+	postEntry.IsNFT = true
+	postEntry.NumNFTCopies = txMeta.NumCopies
+	postEntry.HasUnlockable = txMeta.HasUnlockable
+	postEntry.NFTRoyaltyToCreatorBasisPoints = txMeta.NFTRoyaltyToCreatorBasisPoints
+	postEntry.NFTRoyaltyToCreatorCoinBasisPoints = txMeta.NFTRoyaltyToCreatorBasisPoints
+	bav._setPostEntryMappings(postEntry)
+
+	// Add the appropriate NFT entries.
+	posterPKID := bav.GetPKIDForPublicKey(postEntry.PosterPublicKey)
+	for ii := uint64(1); ii <= txMeta.NumCopies; ii++ {
+		nftEntry := &NFTEntry{
+			OwnerPKID:         posterPKID.PKID,
+			NFTPostHash:       txMeta.NFTPostHash,
+			SerialNumber:      ii,
+			IsForSale:         txMeta.IsForSale,
+			MinBidAmountNanos: txMeta.MinBidAmountNanos,
+		}
+		bav._setNFTEntryMappings(nftEntry)
+	}
+
+	// Add an operation to the utxoOps list indicating we've created an NFT.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:          OperationTypeCreateNFT,
+		PrevPostEntry: prevPostEntry,
+	})
+
+	return totalInput, totalOutput, utxoOpsForTxn, nil
+}
+
+func (bav *UtxoView) _connectUpdateNFT(
+	txn *MsgBitCloutTxn, txHash *BlockHash, blockHeight uint32, verifySignatures bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+
+	// Check that the transaction has the right TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeUpdateNFT {
+		return 0, 0, nil, fmt.Errorf("_connectUpdateNFT: called with bad TxnType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+	txMeta := txn.TxnMeta.(*UpdateNFTMetadata)
+
+	// Verify the NFT entry exists.
+	nftKey := MakeNFTKey(txMeta.NFTPostHash, txMeta.SerialNumber)
+	prevNFTEntry := bav.GetNFTEntryForNFTKey(&nftKey)
+	if prevNFTEntry == nil || prevNFTEntry.isDeleted {
+		return 0, 0, nil, RuleErrorCannotUpdateNonExistentNFT
+	}
+
+	// Verify that the updater is the owner of the NFT.
+	updaterPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
```

Comment:
```suggestion
	updaterPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
	if updaterPKID == nil || updaterPKID.isDeleted {
		return 0, 0, nil, errors.Wrapf(err, "_connectUpdateNFT: ")
	}
```

Reply:
Same comment as above, I'm not sure what "err" is being wrapped here.

Reply:
yurp - should do a fmt.Errorf thing here. a nil or deleted updaterPKID should be an error.